### PR TITLE
Data completeness checker module

### PR DESCRIPTION
**Pull request by npawar**
This pull request contains the data completeness checker code. The data completeness checker will check periodically for completeness of time buckets of datasets.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/completeness/checker/DataCompletenessTaskUtils.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+      ResultSetGroup resultSetGroup = CACHE_REGISTRY.getResultSetGroupCache().get(new PinotQuery(pql, dataset));
+      ResultSet resultSet = resultSetGroup.getResultSet(0);
+      for (int i = 0; i < resultSet.getRowCount(); i++) {
+        Long timeValue = Long.valueOf(resultSet.getGroupKeyString(i, 0));
+        Long count = resultSet.getLong(i, 0);
+        timeValueToCount.put(timeValue, count);
+      }
+
+    } catch (ExecutionException e) {
+      LOG.error("Exception in getting count *", e);
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/completeness/checker/DataCompletenessTaskUtils.java]: Minor suggestion: Put more information to the log, e.g., "Exception in getting count * for dataset {}..."

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/completeness/checker/Wo4WAvgDataCompletenessAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        Long bucketValueMS = entry.getValue();
+        Long baselineBucketValueMS = bucketValueMS - baselineOffset;
+        String baselineBucketName = dateTimeFormatter.print(baselineBucketValueMS);
+        DataCompletenessConfigDTO configDTO = dataCompletenessConfigDAO.findByDatasetAndDateSDF(dataset, baselineBucketName);
+        if (configDTO == null) {
+          baselineBucketNameToBucketValueMS.put(baselineBucketName, baselineBucketValueMS);
+        }
+      }
+      // for all baseline values not present in database, fetch their counts, and update in database
+      LOG.info("Missing baseline buckets {}", baselineBucketNameToBucketValueMS.keySet());
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/completeness/checker/Wo4WAvgDataCompletenessAlgorithm.java]: Add more information such as dataset?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DataCompletenessResource.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    double percentCompleteness = 0;
+    double baselineTotalCount = 0;
+    switch (algorithm) {
+      case WO4W_AVERAGE:
+      default:
+        for (Long baseline : baselineCounts) {
+          baselineTotalCount = baselineTotalCount + baseline;
+        }
+        baselineTotalCount = baselineTotalCount/baselineCounts.size();
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DataCompletenessResource.java]: Could dividing by zero happen here?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/completeness/checker/Wo4WAvgDataCompletenessAlgorithm.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    dataCompletenessResource = new DataCompletenessResource();
+    dataCompletenessConfigDAO = DAO_REGISTRY.getDataCompletenessConfigDAO();
+  }
+
+
+  @Override
+  public void computeBaselineCountsIfNotPresent(String dataset, Map<String, Long> bucketNameToBucketValueMS,
+      DateTimeFormatter dateTimeFormatter, TimeSpec timeSpec) {
+
+    long weekInMillis = TimeUnit.MILLISECONDS.convert(7, TimeUnit.DAYS);
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/completeness/checker/Wo4WAvgDataCompletenessAlgorithm.java]: Could this method encounter the issue of daylight saving time?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DataCompletenessResource.java (hunk 3 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DataCompletenessResource.java]: Shouldn't happen in the code, but could happen if invoked as an endpoint with bad inputs. Handled this

**Pull request merged**